### PR TITLE
Web3 Module API Documentation

### PR DESCRIPTION
**Pull request by nivida**
## Description

Adding of a detailed Web3 Module API Documentation.

Steps before this documentation can be merged: 

- [x] Document all ES6 classes regarding the Web3 Module API 
- [x] Link the classes to the source code until a class reference exist.

## Type of change

- [x] Documentation

### CODE REVIEW

**Comment by Levino** [comment]
Hm..

**Comment by nivida** [comment]
@Levino If you find the time could you please check this documenation if it's understandable for you?

**Comment by levino** [comment]
First feedback: I would create an issue for this and reference here with "Fixes #1928282828".
Please update with base branch.

I will continue to review.

**Comment by levino** [comment]
- Build does not pass.
- Why is there no preview link? How are the docs hosted?

**Comment by levino** [comment]
There is no documentation on how to build the docs locally. There is no hot watcher is there? Created https://github.com/ethereum/web3.js/issues/2250

**Comment by Levino** [comment]
Created https://github.com/ethereum/web3.js/issues/2250

**Comment by levino** [comment]
How do you want me to review this? Checkout the branch, run `npm run docs` and then go through the PDFs? If https://github.com/ethereum/web3.js/issues/2251 was implemented this would be much nicer.

File: packages/web3-core-method/lib/methods/AbstractMethod.js
Comment (levino) [inline packages/web3-core-method/lib/methods/AbstractMethod.js]: I do not see a test changed for this. This is troubling. Either there is no test or the test does not test anything. How could it have worked before? Also I consider this a mix of concerns: I thought this PR is about the docs, now I see code changes. I think the code changes should be moved to another PR.

**Comment by levino** [comment]
I cannot build this locally.

```bash
npm run docs

> web3@1.0.0-beta.38 docs /home/levin/coding/hardfork/web3.js
> cd docs; make html;

sphinx-build -b html -d _build/doctrees   . _build/html
make: sphinx-build: Command not found
Makefile:53: recipe for target 'html' failed
make: *** [html] Error 127
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! web3@1.0.0-beta.38 docs: `cd docs; make html;`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the web3@1.0.0-beta.38 docs script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/levin/.npm/_logs/2019-01-28T11_28_10_130Z-debug.log
```

So one needs to globally install `sphinx-build` on the local machine? Kind of a bad setup. Is it not possible to install this via npm?

**Review by levino — CHANGES_REQUESTED** [review]

**Comment by levino** [comment]
So I will be waiting for you to get back on my comments / questions. Please at least add a "how to run locally".

**Comment by nivida** [comment]
Thanks for your feedback. Yes, I know the documentation setup is not good. The goal of this PR is to add an additional section to the documentation, any other improvements will be done later. The focus currently is on bugs and to add the Web3 Moudle API documentation.

You do not have to checkout the branch for reviewing the content of the documentation. It's also possible to view them directly on [github](https://github.com/ethereum/web3.js/blob/95087d6b1977f9ff07824d083b3c457f00ec00b0/docs/include_web3-module-abstract-method-class-reference.rst).

**Comment by levino** [comment]
So it is just markdown.

**Comment by levino** [comment]
I think it would be much easier if you created one PR for each point in your list above. Having so many issues in one PR is very confusing.

**Comment by levino** [comment]
I do not understand what the point of this PR should be. Could you please verbosely explain the issue that you are trying to solve?

Before merging this PR: Who wants to do what?

Then I can see whether or not that person achieves what they wanted to do after this PR has been merged.

**Comment by nivida** [comment]
I only asked if you could read the documentation and tell me if you understand the content of it without any additional information. The point of this PR is to document the Web3 Module API as the title already says. 

If you find the time could you quickly read through the document and tell me if you are able to create a module with the provided information or not?

**Comment by levino** [comment]
What is a module and why would I create one? I really cannot follow. It is very hard because this is not embedded in the docs atm and I lack context.

**Comment by OFRBG** [comment]
@Levino @nivida I'd like to chime in. I'm highly interested in making web3js documentation as readable as possible. Would it be okay if I propose a specific structure for each topic? I've been using the docs for almost two years now, and I'm relatively familiar with the problems it has.

**Comment by levino** [comment]
Please start by creating issues that describe the problems you encounter in a very understandable (I am a little slow it seems so please consider this) manner. If everyone agrees that these issues are worth fixing, you can start on the fixes.

**Comment by yxliang01** [comment]
I would say the `extend` method should have been retained until this documentation is ready... This transition makes life of developers extending web3 difficult... Maybe it's even better if the Web3 doc allows us to view the doc of older 1.0 beta versions.

**Comment by fixanoid** [comment]
> I would say the `extend` method should have been retained until this documentation is ready... This transition makes life of developers extending web3 difficult... Maybe it's even better if the Web3 doc allows us to view the doc of older 1.0 beta versions.

And I would add that removing this breaks a ton of modules already in use. And while this is helpful in the longer run, it leaves many a module broken until maintainers update them or manual dependency resolution is made to the last working version. Why not leave it available and throw warnings for a while?

**Comment by nivida** [comment]
@fixanoid 
> And I would add that removing this breaks a ton of modules already in use

I've removed this possibility now a month ago and explained why I removed this. Feel free to create a PR if there is a missing JSON-RPC method.

> And while this is helpful in the longer run, it leaves many a module broken until maintainers update them or manual dependency resolution is made to the last working version. 

If you would like to use the ``extend`` method then please use the version beta.37.

> Why not leave it available and throw warnings for a while?

I didn't add this during the refactoring because it's really ugly and hard to test. There are also some problems with the formatters and utils modules which the ``extend`` method was indirectly depending on. 

All the details about the refactoring are explained in the release announcement of version beta.38: 
https://medium.com/@samuel_91690/web3-js-v1-0-0-beta-38-8d2bb6e73d0b

**Pull request merged**
